cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_topology_generator.py

```diff
@@ -0,0 +1,170 @@
+# See the License for the specific language governing permissions and limitations under the License.
+
+# FIXME: Fix Code Duplication
+# pylint: disable=R0801
+
+import argparse
+import logging
+import os
+import traceback
+
+import yaml
+
+log = logging.getLogger()
+
+
+CAPACITY_TYPE_MAP = {
+    "ONDEMAND": "on-demand",
+    "SPOT": "spot",
+    "CAPACITY_BLOCK": "capacity-block",
+}
+CONFIG_HEADER = "# This file is automatically generated by pcluster\n"
+
+
+class CriticalError(Exception):
+    """Critical error for the script."""
+
+    pass
+
+
+class ConfigurationFieldNotFoundError(Exception):
+    """Field not found in configuration."""
+
+    pass
+
+
+def _load_cluster_config(input_file_path):
+    """Load cluster config file."""
+    with open(input_file_path, encoding="utf-8") as input_file:
+        return yaml.load(input_file, Loader=yaml.SafeLoader)
+
+
+def generate_topology_config_file(output_file: str, input_file: str, block_sizes: str):  # noqa: C901
+    """
+    Generate Topology configuration file.
+
+    Generate topology.conf
+
+    # This file is automatically generated by pcluster
+    BlockName=block1 Nodes=queue-1-st-compute-resource-0-[1-9]  #### 9 nodes
+    BlockName=block2 Nodes=queue-1-st-compute-resource-0-[1-18] #### 18 nodes
+    BlockSizes=9,18
+    """
+    if block_sizes:
+        min_block_size_list = min(list(map(int, block_sizes.split(","))))
+        max_block_size_list = max(list(map(int, block_sizes.split(","))))
+
+        cluster_config = _load_cluster_config(input_file)
+        queue_name, compute_resource_name = None, None
+        try:
+            topology_config = CONFIG_HEADER + "\n"
+            block_count = 0
+            for queue_config in cluster_config["Scheduling"]["SlurmQueues"]:
+                queue_name = queue_config["Name"]
+
+                # Retrieve capacity info from the queue_name, if there
+                queue_capacity_type = CAPACITY_TYPE_MAP.get(queue_config.get("CapacityType", "ONDEMAND"))
+                if queue_capacity_type != CAPACITY_TYPE_MAP.get("CAPACITY_BLOCK"):
+                    log.info("ParallelCluster does not create topology for %s", queue_capacity_type)
+                    continue
+
+                for compute_resource_config in queue_config["ComputeResources"]:
+                    compute_resource_name = compute_resource_config["Name"]
+                    compute_min_count = compute_resource_config["MinCount"]
+                    compute_max_count = compute_resource_config["MaxCount"]
+                    if compute_min_count == compute_max_count:
+                        node_type = "st"
+                    else:
+                        continue
+
+                    # Check for if reservation is for NVLink and size matches min_block_size_list
+                    if compute_resource_config.get("InstanceType") == "p6e-gb200.36xlarge":
```

Comment:
Why just this instance type? .72xlarge? Any way to more programatically determine these?